pynxtools/test_suite/README.md

```diff
@@ -0,0 +1,100 @@
+ |---pyproject.toml
+ |---test
+       |---data/<test_data>
+       |---test_plugin.py
+```
+
+This hierarchical structure allows `pynxtools` to set up the integration for plugin test from its own test script. Plugin can add multiple example for multiple version and type of the raw data files from an experiment techniques. Plugin developers can wish to put other raw files related to the plugin owned test as they want (though usually they go to test/data/ dir). The `examples` folder must have one or multiple sub-directory where each directory represents a single example of input files (e.g. eln.yaml, raw_file.ext) to launch the plugin reader and a output file of `.nxs` extension genrated by reader from the given banch of input files. Beside the input files, a `test_config.json` file must be defined with the information of `nxdl` (e.g. `NXsts`, `NXmpes`), `reader` (e.g. `STMReader`, `MPESReader`), `plugin_name` (e.g. `pynxtools-stm`, `pynxtools-mpes`) and `example_dir` (e.g. `*`, `*_1`, `example_1`). Therfore, the `test_config.json` gives the content for reader where the test data and other infos to launch the reader. Note that to define the value of the `exmaple_dir` unix style pathname pattern expansion (also used in [glob](https://docs.python.org/3/library/glob.html) lib), still we recomand to use simply the name of the example directory. A few examples forn `test_condig.json` has been added bellow.
+
+The `test_config.json` can be defined to run on all available example.
+
+```json
+# test_condig.json
+{"launch_data":
+    [
+        {"nxdl": "NXfoo",
+        "reader": "FOOReader",
+        "plugin_name": "pynxtools-FOO",
+        "example_dir": "*"}
+    ]
+}
+```
+The `test_config.json` file can be expanded as
+```json
+# test_condig.json
+{"launch_data":
+    [
+        {"nxdl": "NXfoo",
+        "reader": "FOOReader",
+        "plugin_name": "pynxtools-FOO",
+        "example_dir": "example_1"},
+        {"nxdl": "NXfoo",
+        "reader": "FOOReader",
+        "plugin_name": "pynxtools-FOO",
+        "example_dir": "example_2"}
+    ]
+}
+```
+**Note** : There might be a case, a specific `exmaple` directory might be intended for a specific `nxdl`. In that case the value of the `nxdl` must be the desired name of `nxdl`.
+
+## How to write integration test from the reader plugin
+
+It is very simple to write a test to verify the plugin integration with `pynxtools` from plugin test. The developer can write the test where they want, only they need to use the provided test interface from `pynxtools`. An example test for `pynxtools-FOO` (a demo) plugin as follows:
+
+```python
+# test_plugin.py
+
+import os
+
+from pynxtools_foo.reader import FOOReader
+import pytest
+from pynxtools.test_suite.reader_plugin import ReaderTest
+
+module_dir = os.path.dirname(os.path.abspath(__file__))
+
+
+@pytest.mark.parametrize(
+    "nxdl,reader,files_or_dir",
+    [
+        ("NXfoo", FOOReader, f"{module_dir}/../examples/example_1"),
```

Comment:
Example data can come as it was before i.e. in test/data/, This is the integration test with pynxtools, to check any changes of the pynxtools or plugin break the work flow, not to be copulsary all the examples should in /examples/example_1 ... . So, a few light examples can be added in the `examples` dir. Even in this example script I added two examples in the format `examples/example_1` `examples/example_2` and one from `/tests/data/example_3`.

Reply:
I agree with @lukaspie that data which is used for tests should not live in the examples directory. In my view examples might be added but I think it's generally better if they have some kind of tutorial fashion, i.e., providing jupyter notebooks people can follow through and not just plain data files.

Reply:
> Example data can come as it was before i.e. in test/data

My whole point is that example data IS NOT test data. They should be separated and thus the test data should live within the tests folder. You can (and probably should) always supply examples as well, but these should not be required for the tests.

My suggestions is that the tests script (both in the individual plugins and also the one proposed here for pynxtools) calls the `tests/data` folder of the plugin.

Reply:
@lukaspie, @domna, The example files is for the test purpose of the reader from pynxtools. Usually, the tests dir does not comes with packages thereofore need to place a few example package leve. we could rethink about the place where a few examples should go and they must come with the reader packages for testing in pynxtools. If anyone of you see more reasonable spot for the example files, I am keen to change my mind.

Reply:
@lukaspie, without input files there are no ways to run the tests for diverse reader plugin. Definitely, the test data should come with the plugin package, then the `pynxtools` would pick the correspoding test files and run the plugin test.

Reply:
Okay, now I at least understand what you are saying. But then the example files (as you envision it) should be within the pynxtools-FOO module, otherwise they don't come with the package as well, right? So, with the structure you have in the readme here, i.e., examples, pynxtools-FOO, and tests on the same level, neither the examples nor the tests get shipped with the package, right? In that case, the example files would also not be available for testing in pynxtools (where we install the plugins before the test).

Reply:
But how would you ship the examples with the plugin if they are at the same level. The suggested structure is
```
pynxtools-FOO
 |---examples
 |---pynxtools_FOO # this is the package directory
 |     |---<pynxtools-FOO package stuffs>
 |---pyproject.toml
 |---tests
 |---test_config.json
```
That is, `examples` and `test_config` are outside of the package dir. How would it be possible to ship `examples` and `test_config` together with the package? I.e., what do I have to write into my project.toml file of the plugin?

Reply:
We should not just put the files there for the sake of including them in the package for testing them here. The `test` dir and testfiles are excluded for a reason and they  _should not_ be included in the package which is available for users of the package. Even here it will be very problematic, because _all_ plugins will create a folder named examples in the `site-packages`, which then will be overwritten by the next plugin which gets installed. We should rather think about a mechanism here to get the example/test data from the packages without including unnecessary files in the package distribution, i.e., cloning them and `pip install -e` could be a possibility. But generally, testing files outside the standard distribution is a tricky business.

Reply:
Why don't we just clone all the repos in the ci/cd and run the tests independently in each of the readers?
This is much simpler than going through the python dependency chain.

Reply:
Thanks! @domna and @lukaspie, here I got some comments that I need to work on. 

But, ci/cd is failing because the plugin did not include any plugin in the python installation command. So, no plugging has been found and the test is running on the pynxtools standalone.

Reply:
Ok perfect. Let me know whenever you have a new version of this PR available, so that I can update the corresponding test and data structure in pynxtools.

Reply:
> Why don't we just clone all the repos in the ci/cd and run the tests independently in each of the readers? This is much simpler than going through the python dependency chain.
> 
> Edit: But this might be problematic for local testing I just realized... but maybe that's fine? I let ci/cd do the tests anyways most of the time

Initially, this comment I did not understand. I am not 100% agree with the cloning of the repo in ci/cd. One very good reason you already mentioned the issue for the local test. 

I am not comfortable pushing something and waiting for the result of the test for up to an hour (considering the desperately heavy plugin) to check all the plugins. 

Still, I and wondering, Even cloning the repo where to get the test script of the plugin package? By default, the plugin package does not contain any test script or data.

Reply:
> Still, I and wondering, Even cloning the repo where to get the test script of the plugin package? By default, the plugin package does not contain any test script or data.

There is a [test script](https://github.com/FAIRmat-NFDI/pynxtools-plugin-template/blob/main/pynxtools-%7B%7Bcookiecutter.reader_name%7D%7D/tests/test_reader.py) already in the cookiecutter template, so any plugin created from there should have tests and test data.

I would rather loose the ability to locally test instead of the route through the examples. As Florian pointed out, that doesn't work at all since the site-packages/examples will be overwritten.

Reply:
Maybe we should also look into more modern testing/packaging packages, i.e., tox, poetry or hatch (this is not a complete list, just the one which came quickly to my mind). They might already have proper solutions for our problems.

Reply:
I agree with Florian, the example with_overwite the all the other example from other plugin. 
One way to resolve this is to append a part denoting the name of plugin such as `examples_xps`. Which is nice and clearer denotation for the examples.

Reply:
Right, locally it would take almost the same time as it would take in ci/cd. But I meant tha one does not need to run all the tests to check a single line of code, locally it is also possible to run a single test separately.

Reply:
> > I am not comfortable pushing something and waiting for the result of the test for up to an hour (considering the desperately heavy plugin) to check all the plugins.
> 
> So far our tests are rather in the time range of ~5 - 10 mins and local testing normally isn't much faster. So if your tests take an hour, you'll also need to wait locally for an hour. However, we do parallel tests in the ci/cd for multiple versions so until _everything_ finishes might take a while longer, but normally seeing on test pass or fail is enough to already work on changes.
> 
> > Still, I and wondering, Even cloning the repo where to get the test script of the plugin package? By default, the plugin package does not contain any test script or data.
> 
> If you clone the repo everything is included. The packaging strips down everything which isn't explicitly included to not bloat the package with tests, examples and documentation (which cannot be used and or viewed for installed packages), but cloning is not running any packaging routines. Only if you run `pip install .` the directory is packaged. If you use `pip install -e .`, also the other folders should be available. But even there referencing is partly undefined, this is why I created the issue for the `src` layout #300.
> 
> One additional consideration on what we could do is specifying test/example files with a doi (e.g., from zenodo or nomad) and then download them and run the tests against them. This is something we already discussed with @mkuehbach, who wants to have a mechanism to tests against many and large files, which he doesn't want to include in any repository. This can be implemented in python and easily run in python, but we could also do it as a local script and run the test only if the folder exists. So you can decide if you want to bloat your local install with all these files.

Including test feature can run tests on the data file from `zenodo` and it is a good idea. But, I strongly do not want to add this feature with PR which was not planned here. This type of test would go to the plugin's tests compartment, not to the `pynxools`. The main purpose to set an on-call test in `pynxtools` for each plugin (obviously with light weight data set), if any unintentional change in pynxtools breaks any plugin.

Reply:
> Maybe we should also look into more modern testing/packaging packages, i.e., tox, poetry or hatch (this is not a complete list, just the one which came quickly to my mind). They might already have proper solutions for our problems.

Thanks for pointing that out. One way to include the test data into the examples folder is to add a `MANIFEST.in` file next to the `pyproject.toml`. With this setup the example needs to got into `pynxtools_foo` module dir.

Reply:
I agree. This are two different types of files. However, if this are just stripped down versions for simple testing I would even _stronger_ argue against putting them into a folder named `examples`. This is very confusing for people navigating to the repo. They will expect to see some example data but they will just find some stripped down testing files which in many cases is not useful to them at all.

I still believe that checking out the reader plugins at their respective version is the best option but I could also envision alternatives where we adapt the packaging.
But I believe it's not worth the effort to only enable quicker local tests. It's way easier to just have a clone of the plugin reader, install the respective pynx version in the venv and then run the tests if one wants to test a particular function. For any unexpected behaviour we would have it in ci/cd (and in that case you would also not test a particular function).

Reply:
> I agree. This are two different types of files. However, if this are just stripped down versions for simple testing I would even _stronger_ argue against putting them into a folder named `examples`. This is very confusing for people navigating to the repo. They will expect to see some example data but they will just find some stripped down testing files which in many cases is not useful to them at all.
> 
> I still believe that checking out the reader plugins at their respective version is the best option but I could also envision alternatives where we adapt the packaging. But I believe it's not worth the effort to only enable quicker local tests. It's way easier to just have a clone of the plugin reader, install the respective pynx version in the venv and then run the tests if one wants to test a particular function. For any unexpected behaviour we would have it in ci/cd (and in that case you would also not test a particular function).

Yesterday I did not see this conversation. Now I understand, @sanbrock point on today's meetings for name of the `examples` and running some test fundtions. The name came immediately to my mind while creating the example folder.

@domna Cloning a repo, installing the repo, installing the pynx is also good idea and really an extensive test on the plugin. In this case, the intended tests are to test different types and versions of files probably with different nxdl definitions, which will verify the plugin's compatibility with all the file readers the plugin contains. 

Regarding the many different file types and versions in a plugin, I am very skeptical the pynx will take a huge time to finish the ci/cd (which was the case for Nomad ci/cd, and developers complained about such elongated time for ci/cd).

I must admit that it was not in my plan. if @domna and @lukaspie both of you (as both of you taking interactive part in this PR) agree, I guess the approach cloning the repo is still doable.